File: server/task/fixPresite.go
package task

import (
	"fmt"
	"io"
	"log"
	"os"
	"regexp"
	"strings"
)

// Closing content of the prerendered file. presite doesn't include a trailing newline!
const closingContent = `</body></html>`

// Yep, using a regex to find an HTML tag
var matchScriptSrc = regexp.MustCompile(`<script src="(/static/js/[^j]*js)">`)

func handleRewritePresiteIndexTask(buildIndexPath string, presiteIndexPath string) error {
	baseFile, err := os.Open(buildIndexPath)
	if err != nil {
		return fmt.Errorf("could not open base file %s: %w", buildIndexPath, err)
	}
	defer baseFile.Close()
	baseFileBytes, err := io.ReadAll(baseFile)
	if err != nil {
		return fmt.Errorf("could not read all from base file %s: %w", buildIndexPath, err)
	}
	log.Printf("Read build file %s", buildIndexPath)

	matches := matchScriptSrc.FindAll(baseFileBytes, -1)
	if len(matches) < 3 {
		return fmt.Errorf("Expected to get at least 3 script tags, got %v", matches)
	}
	var result strings.Builder
	for _, match := range matches {
		log.Printf("Found `%s`", match)
		result.Write(match)             // whole <script> tag
		result.WriteString("</script>") // and closing tag
	}

	presiteFile, err := os.OpenFile(presiteIndexPath, os.O_RDWR, 0o755)
	if err != nil {
		return fmt.Errorf("could not open presite file %s: %w", presiteIndexPath, err)
	}
	defer presiteFile.Close()
	log.Printf("Opened presite file %s", presiteIndexPath)
	info, err := presiteFile.Stat()
	if err != nil {
		return fmt.Errorf("could not stat prestie file %s: %w", presiteIndexPath, err)
	}
	offset := info.Size() - int64(len(closingContent))
	if _, err = presiteFile.Seek(offset, 0); err != nil {
		return fmt.Errorf("could not seek to offset %v (of %v) in %s: %w", offset, info.Size(), presiteIndexPath, err)
	}
	log.Printf("Seeked to end of file")

	if _, err = presiteFile.WriteString(result.String()); err != nil {
		return fmt.Errorf("could not write to presite file %s: %w", presiteIndexPath, err)
	}
	log.Printf("Wrote all the tags in")
	if _, err = presiteFile.WriteString(closingContent); err != nil {
		return fmt.Errorf("could not write closing to presite file %s: %w", presiteIndexPath, err)
	}
	return nil
}

File: web/script/fix-presite.ts
import * as process from 'process';
import * as fsBase from 'fs';
import * as fs from 'fs/promises';
import * as path from 'path';
import * as scriptUtil from './scriptUtil';

//
// About this script
////////////////////
//
// At time of writing, presite on my machine is not including <script> tags that
// point to the build project's script sources. This script finds and injects them.
//
// This has not been tested with code splitting. This script should be run for each
// entrypoint.

// closingContent is the ending content of the file. Presite doesn't include a
// trailing newline.
const closingContent = `</body></html>`;

// matchScriptSrc is indeed using a regular expression to parse HTML.
const matchScriptSrc = /<script src="(\/static\/js\/[^j]*js)">/g;


// Check for invalid invocation
if (process.argv.length < 3) {
    console.error(process.argv);
    console.error("Usage: fix-presite.ts <build index path> <presite build index path>");
    process.exit(1);
}

// Get paths from reading cmd args - in this case we have a lot of leading args for
// tsnode, tsnode flags, and filename, so we just grab the last two args.
const buildIndexPath = process.argv[process.argv.length - 2];
const presiteIndexPath = process.argv[process.argv.length - 1];

// Read the base build file's content.
const baseContent = fsBase.readFileSync(buildIndexPath, { encoding: "utf8" });

// Find any <script> tags in the base file.
const matches = baseContent.matchAll(matchScriptSrc);

// Create a buffer of <script src='src.js'/></script> for the scripts
const scriptTags = [];
for (const [tag, src] of matches) {
    console.log("Found ", src);
    scriptTags.push(tag);
    scriptTags.push("</script>");
}
const scriptTagsText = scriptTags.join("");
const scriptTagsFound = scriptTags.length / 2;

// We should have at least 3 scripts: runtime, deps, index.
if (scriptTagsFound < 3) {
    const paths = [];
    for (const [_, src] of matches) {
        paths.push(src);
    }
    throw new Error(
        `Expected to get at least 3 script tags, got ${scriptTagsFound}: ${paths.join(", ")}`
    );
}

// Open output file
fs.open(presiteIndexPath, fsBase.constants.O_RDWR, 0o755)
    // Get stats to calculate offset
    .then(handle => {
        console.log("Opened", presiteIndexPath);
        return Promise.all([handle, handle.stat()]);
    // Write result into file at offset
    }).then(([handle, stats]) => {
        console.log("Found", path.basename(presiteIndexPath), ":", scriptUtil.bytesToMetric(stats.size));
        const offset = stats.size - closingContent.length;

        return Promise.all([handle, handle.write(scriptTagsText, offset)]);
    // Check results, write closing content
    }).then(([handle, result]) => {
        if (result.bytesWritten != scriptTagsText.length) {
            throw new Error(`Wrote ${result.bytesWritten} of ${scriptTags.length} bytes`);
        }
        console.log("Wrote", scriptTagsFound, "script references to file");
        return Promise.all([handle, handle.write(closingContent)]);
    // Check results, close file
    }).then(([handle, result]) => {
        if (result.bytesWritten != closingContent.length) {
            throw new Error(`Wrote ${result.bytesWritten} of ${scriptTags.length} bytes`);
        }
        return handle.close();
    // Assert we closed the file.
    }).then(() => {
        console.log("Build complete.");
    }).catch(err => {
        console.error("Unable to write file:", err);
    });
